File: visualmserver/src/main/java/visualmserver/models/Material.java
package visualmserver.models;

import org.hibernate.annotations.CreationTimestamp;

import javax.persistence.*;
import javax.validation.constraints.NotBlank;
import javax.validation.constraints.NotNull;
import javax.validation.constraints.Size;
import java.time.LocalDate;
import java.util.List;
import java.util.Set;

@Entity
@Table(name = "material")
public class Material {
    public static final int MAXIMUM_STEPS = 8;
    public static final int MAXIMUM_INGREDIENTS = 6;

    @NotNull
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    @Column(name = "sequence_number")
    private Long sequenceNumber;

    @Column(name = "sequence_number_published")
    private Long sequenceNumberPublished;

    @Size(max = 25)
    @NotBlank(message = "Name cannot be empty.")
    private String name;

    @NotNull
    @CreationTimestamp
    @Column(name = "creation_date")
    private LocalDate creationDate;

    @NotNull
    @NotBlank
    @Column(columnDefinition = "TEXT")
    private String steps;

    @NotNull
    @NotBlank
    @Size(max = 200)
    @Column(columnDefinition = "TEXT")
    private String changes;

    @NotNull
    @NotBlank
    private String reference;

    @Column(name = "overview_url")
    private String overviewURL;

    @Column(name = "close_up_url")
    private String closeUpURL;

    @NotNull
    @NotBlank
    @Column(name = "qrcode_url")
    private String qrCodeURL;

    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(name = "material_has_tag",
            joinColumns = @JoinColumn(name = "sequence_number"),
            inverseJoinColumns = @JoinColumn(name = "id")
    )
    private List<Tag> tags;

    @NotNull
    @OneToMany(mappedBy = "material", cascade = {CascadeType.REMOVE})
    private Set<MaterialIngredient> materialIngredients;

    @NotNull
    @Column(name = "status")
    @Enumerated(EnumType.STRING)
    private SaveStatus saveStatus;

    @Enumerated(EnumType.STRING)
    private MaterialType type;

    @ManyToOne
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    @Column(name = "parent_sequence_number")
    private Long parentId;

    public Material() {
    }

    public Long getSequenceNumber() {
        return this.sequenceNumber;
    }

    public Long getSequenceNumberPublished() {
        return this.sequenceNumberPublished;
    }

    public void setSequenceNumber(Long sequenceNumber) {
        this.sequenceNumber = sequenceNumber;
    }

    public void setSequenceNumberPublished(Long sequenceNumberPublished) {
        this.sequenceNumberPublished = sequenceNumberPublished;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public LocalDate getCreationDate() {
        return creationDate;
    }

    public void setCreationDate(LocalDate creationDate) {
        this.creationDate = creationDate;
    }

    public String getSteps() {
        return steps;
    }

    public void setSteps(String steps) {
        this.steps = steps;
    }

    public String getChanges() {
        return changes;
    }

    public void setChanges(String changes) {
        this.changes = changes;
    }

    public String getOverviewURL() {
        return overviewURL;
    }

    public void setOverviewURL(String mediaURL) {
        this.overviewURL = mediaURL;
    }

    public String getQrCodeURL() {
        return qrCodeURL;
    }

    public void setQrCodeURL(String qrCodeURL) {
        this.qrCodeURL = qrCodeURL;
    }

    public List<Tag> getTags() {
        return tags;
    }

    public void addTag(Tag tag) {
        this.tags.add(tag);
        tag.addMaterial(this);
    }

    public SaveStatus getSaveStatus() {
        return saveStatus;
    }

    public void setSaveStatus(SaveStatus saveStatus) {
        this.saveStatus = saveStatus;
    }

    public MaterialType getType() {
        return type;
    }

    public void setType(MaterialType type) {
        this.type = type;
    }

    public Set<MaterialIngredient> getMaterialIngredients() {
        return materialIngredients;
    }

    public void setMaterialIngredients(Set<MaterialIngredient> materialIngredients) {
        this.materialIngredients = materialIngredients;
    }

    public User getUser() {
        return user;
    }

    public void setUser(User user) {
        this.user = user;
    }

    public Long getParentId() {
        return parentId;
    }

    public void setParentId(Long parentId) {
        this.parentId = parentId;
    }

    public String getReference() {
        return reference;
    }

    public void setReference(String reference) {
        this.reference = reference;
    }

    public String getCloseUpURL() {
        return closeUpURL;
    }

    public void setCloseUpURL(String closeUpURL) {
        this.closeUpURL = closeUpURL;
    }
}

File: visualmfrontend/src/app/models/material.ts
import {MaterialIngredient} from './material-ingredient';
import {MaterialType} from './material-type.enum';
import {SaveStatus} from './save-status.enum';
import {Tag} from './tag';
import {User} from './user';

export class Material {
  public static readonly MAXIMUM_STEPS = 10;
  public static readonly MAXIMUM_INGREDIENTS = 6;
  private static FORMAT = '00000';
  private sequenceNumber: number;
  sequenceNumberPublished: number;
  private name: string;
  private creationDate: Date;
  private changes: string;
  private reference: string;
  private steps: string;
  private overviewURL: string;
  private closeUpURL: string;
  private qrCodeURL: string;
  private tags: Tag[];
  private materialIngredients: MaterialIngredient[];
  private saveStatus: SaveStatus;
  private type: MaterialType;
  private user: User;
  private parentId: number;

  constructor(sequenceNumber?: number, name?: string, changes?: string, steps?: string, qrCodeURL?: string,
              tags?: Tag[], materialIngredients?: MaterialIngredient[], saveStatus?: SaveStatus, type?: MaterialType, user?: User,
              parentId?: number, reference?: string) {
    this.sequenceNumber = sequenceNumber;
    this.name = name;
    this.changes = changes;
    this.creationDate = new Date();
    this.steps = steps;
    this.qrCodeURL = qrCodeURL;
    this.tags = tags;
    this.materialIngredients = materialIngredients;
    this.saveStatus = saveStatus;
    this.type = type;
    this.user = user;
    this.parentId = parentId;
    this.reference = reference;
  }

  public getSequenceNumber(): number {
    return this.sequenceNumber;
  }

  public getFormattedSequenceNumber(): string {
    if (this.sequenceNumberPublished == null) {
      return "OLD" + this.sequenceNumber;
    } else {
      return Material.FORMAT.substring(0, Material.FORMAT.length - this.getSequenceNumberPublished().toString().length) + this.getSequenceNumberPublished().toString();
    }
  }

  public getSequenceNumberPublished(): number {
    return this.sequenceNumberPublished;
  }

  public getName(): string {
    return this.name;
  }

  public getSteps(): string {
    return this.steps;
  }

  public setSteps(step): string {
    return this.steps;
  }

  public getCreationDate(): Date {
    return this.creationDate;
  }

  public getTags(): Tag[] {
    return this.tags;
  }

  public setTags(tags: Tag[]): void {
    this.tags = tags;
  }

  public getSaveStatus(): SaveStatus {
    return this.saveStatus;
  }

  public getOverviewURL(): string {
    return this.overviewURL;
  }

  public setOverviewURL(overviewURL: string): void {
    this.overviewURL = overviewURL;
  }

  public getUser(): User {
    return this.user;
  }

  public getQRCodeURL(): string {
    return this.qrCodeURL;
  }

  static trueCopy(material): Material {
    return Object.assign(new Material(), material);
  }

  public getType(): MaterialType {
    return this.type;
  }

  public getMaterialIngredients(): MaterialIngredient[] {
    return this.materialIngredients;
  }

  public setMaterialIngredients(materialIngredients: MaterialIngredient[]): void {
    this.materialIngredients = materialIngredients;
  }

  public setSequenceNumberPublished(sequenceNumberPublished: number): void {
    this.sequenceNumberPublished = sequenceNumberPublished;
  }

  public getChanges(): string {
    return this.changes;
  }

  public getReference(): string {
    return this.reference;
  }

  public getParentId(): number {
    return this.parentId;
  }


  public setParentId(parentId: number): void {
    this.parentId = parentId;
  }

  public getCloseUpURL(): string {
    return this.closeUpURL;
  }

  public setCloseUpURL(closeUpURL: string): void {
    this.closeUpURL = closeUpURL;
  }
}
